src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java

```diff
@@ -1128,6 +1133,123 @@ void verifyJar(String jarName)
+                    .collect(Collectors.toList());
+
+            var cenEntries2 = cenEntries.getFirst().equals(JarFile.MANIFEST_NAME)
+                    ? cenEntries.subList(1, cenEntries.size()) : cenEntries;
+
+            if (!cenEntries2.equals(locEntries)) {
+                crossChkWarnings.add(rb.getString(
+                        "entries.mismatch.when.comparing.jarfile.and.jarinputstream"));
+            }
+        }
+    }
+
+    private void readEntry(InputStream is) throws IOException {
+        is.transferTo(OutputStream.nullOutputStream());
+    }
+
+    private boolean compareManifest(Manifest cenManifest, Manifest locManifest) {
+        boolean validManifest = true;
+
+        if (cenManifest == null) {
+            crossChkWarnings.add(rb.getString(
+                    "manifest.missing.when.reading.jarfile"));
+            return false;
+        }
+        if (locManifest == null) {
+            crossChkWarnings.add(rb.getString(
+                    "manifest.missing.when.reading.jarinputstream"));
+            return false;
+        }
+
+        Attributes cenMainAttrs = cenManifest.getMainAttributes();
+        Attributes locMainAttrs = locManifest.getMainAttributes();
+
+        for (Object key : cenMainAttrs.keySet()) {
+            Object cenValue = cenMainAttrs.get(key);
+            Object locValue = locMainAttrs.get(key);
+
+            if (locValue == null) {
+                crossChkWarnings.add(String.format(rb.getString(
+                        "manifest.attribute.1.present.when.reading.jarfile.but.missing.via.jarinputstream"),
+                        key));
+                validManifest = false;
+            } else if (!cenValue.equals(locValue)) {
+                crossChkWarnings.add(String.format(rb.getString(
+                        "manifest.atrribute.1.differs.jarfile.value.2.jarinputstream.value.3"),
+                        key, cenValue, locValue));
+                validManifest = false;
+            }
+        }
+
+        for (Object key : locMainAttrs.keySet()) {
+            if (!cenMainAttrs.containsKey(key)) {
+                crossChkWarnings.add(String.format(rb.getString(
+                        "manifest.attribute.1.present.when.reading.jarinputstream.but.missing.via.jarfile"),
+                        key));
+                validManifest = false;
+            }
+        }
+
+        return validManifest;
+    }
+
+    private void compareSigners(JarEntry cenEntry, JarEntry locEntry) {
+        CodeSigner[] cenSigners = cenEntry.getCodeSigners();
+        CodeSigner[] locSigners = locEntry.getCodeSigners();
+
+        boolean cenHasSigners = cenSigners != null;
+        boolean locHasSigners = locSigners != null;
+
+        if (cenHasSigners && locHasSigners) {
+            List<CodeSigner> cenSignerList = Arrays.asList(cenSigners);
+            List<CodeSigner> locSignerList = Arrays.asList(locSigners);
+
+            if (!cenSignerList.equals(locSignerList)) {
```

Comment:
I think you can just call `Arrays.equals()` here.

Reply:
Done.